### PR TITLE
Refactor/step child rendering

### PR DESCRIPTION
**Pull request by trezy**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -24,9 +25,19 @@ export class Wizard extends Component {
     Local Properties
   \***************************************************************************/
 
+  static defaultProps = {
```
Comment (UncleClapton) [inline src/index.js]: I usually stick `defaultProps` and `propTypes` down into their own section at the bottom of a class as they serve a specific purpose outside of the functional code.

File: src/index.js [hunk 2]
```diff
@@ -40,13 +51,19 @@ export class Wizard extends Component {
     Public Methods
   \***************************************************************************/
 
-  setStep = (stepId, newState) => {
+  componentDidMount () {
```
Comment (UncleClapton) [inline src/index.js]: componentDidMount shouldn't be used in this case as it only serves to waste update cycles. By the time state is defined, props is already available to us so we can set this by changing the state property to:

```
state = {
  currentStep: this.props.defaultStep || this.steps[0]
  ...
}
```

File: src/index.js [hunk 3]
```diff
@@ -59,25 +76,97 @@ export class Wizard extends Component {
...
+  /***************************************************************************\
+    Public Methods
+  \***************************************************************************/
+  render () {
+    const { id } = this.props
+
+    return (
+      <Consumer>
+        {({ currentStep, ...childProps }) => (
+          <div hidden={currentStep !== id}>
```
Comment (UncleClapton) [inline src/index.js]: I'm still not the biggest fan of preemptively applying this. We're still functionally mounting a component earlier than a user of the library would expect, there are potential performance problems on initial mount, and there is no benefit until we actually get time slicing.

**Review by UncleClapton — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 3 above)
Comment (trezy) [reply src/index.js]: Your face is no benefit.

File: src/index.js (hunk 2 above)
Comment (trezy) [reply src/index.js]: Legit.

File: src/index.js (hunk 1 above)
Comment (trezy) [reply src/index.js]: I would disagree with this. 

1. They serve a similar purpose as other local properties such as `state`.
1. They may not be directly accessed inside the class, but they serve as documentation for `this.props` which *will* be accessed inside the class and receives no documentation anywhere else.

File: src/index.js [hunk 4]
```diff
@@ -161,11 +161,7 @@ export class Step extends Component {
 
     return (
       <Consumer>
-        {({ currentStep, ...childProps }) => (
-          <div hidden={currentStep !== id}>
-            {this._renderChildren(childProps)}
-          </div>
-        )}
+        {({ currentStep, ...childProps }) => this._renderChildren(childProps)}
```
Comment (UncleClapton) [inline src/index.js]: uhm now currentStep isn't being handled at all. make sure do do a conditional render for currentStep

**Review by UncleClapton — CHANGES_REQUESTED** [review]

File: examples/src/index.js [hunk 5]
```diff
@@ -11,13 +11,36 @@ import {
...
 const App = () => (
   <Wizard defaultStep="foo">
-    <Step id="foo">
-      <span>Foo</span>
-    </Step>
-
-    <Step id="bar">
-      <span>Bar</span>
-    </Step>
+    {({
```
Comment (UncleClapton) [inline examples/src/index.js:14]: I think wizard control rendering would be better served as an optional render prop that defaults to render something like this.

File: examples/src/index.js [hunk 6]
```diff
@@ -11,13 +11,36 @@ import {
...
+            <span>Foo</span>
+          </Step>
+
+          <Step id="bar">
+            <span>Bar</span>
+          </Step>
+        </div>
+
+        {(hasPreviousStep) && (
+          <button onClick={previousStep}>
```
Comment (UncleClapton) [inline examples/src/index.js]: you're also missing type="button" here

File: examples/src/index.js [hunk 7]
```diff
@@ -11,13 +11,36 @@ import {
...
+            Previous
+          </button>
+        )}
+
+        {(hasNextStep) && (
+          <button onClick={nextStep}>
```
Comment (UncleClapton) [inline examples/src/index.js]: and here too

File: src/components/Wizard.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      throw new Error(`Attempting to add duplicate step ID: ${stepID}`)
+    }
+
+    const newSteps = [...steps, stepID]
+
+    const wizardHasMoreThanOneStep = newSteps.length > 1
+    const currentStepIsFirstStep = currentStep === newSteps[0]
+    const currentStepIsLastStep = currentStep === newSteps[newSteps.length - 1]
+
+    console.group('Adding step...', stepID)
```
Comment (UncleClapton) [inline src/components/Wizard.js]: all dem console lines tho

File: examples/src/index.js (hunk 5 above)
Comment (trezy) [reply examples/src/index.js:14]: I'm fine with adding that, but I think we need to make it possible for implementations to access the necessary props to handle the controls themselves.

Comment (UncleClapton) [reply examples/src/index.js:14]: Just as long as it's optional. IE: do a check to see if child is function before calling it

File: src/components/Wizard.js (hunk 8 above)
Comment (trezy) [reply src/components/Wizard.js]: Still deeeeeeeeeeebbbbbbbbbbbbuuuuuuuuuuuggggggggggggggggiiiiiiiiiiiiiinnnnnnnnnnnnngggggggggggg

**Pull request merged**